multipass transfer only transfers half the file (every 2nd 32K block)

**Describe the bug**
Building a vagrant box with bartender fails, because only half of the ingredients.tar.gz file is transferred by `multipass transfer`.

**To Reproduce**

Run multipassd as root in one terminal window, and in another run:

```
UBUNTU_SERIES=noble
./ubuntu-bartender --livecd-rootfs-branch ubuntu/$UBUNTU_SERIES --build-provider multipass -- --series $UBUNTU_SERIES --image-target vagrant --project ubuntu-cpc
```
It fails (see below). Then:

```
debug_bartender_name=(name from ubuntu-bartender output above)
multipass transfer "ingredients.tar.gz" "$debug_bartender_name:ingredients.tar.gz"
multipass transfer "$debug_bartender_name:ingredients.tar.gz" "bad_ingredients.tar.gz"
diff ingredients.tar.gz bad_ingredients.tar.gz 
```

**Expected behavior**
No output (the files are the same).

**Actual behavior**
```
Binary files ingredients.tar.gz and bad_ingredients.tar.gz differ
```

**Logs**
I dit not get any logs or output from multipassd despite running with `-V debug`. It only produced output when terminated with Ctrl+C.

**Additional info**
 - OS: Linux Mint 22
- `multipass version`: 1.16.0-dev.134+g637e43139
- `multipass info`:
```
Name:           noted-ant-ubuntu-bartender
State:          Running
Snapshots:      0
IPv4:           10.196.80.87
Release:        **Ubuntu 20.04.6 LTS**
Image hash:     5c054e515add (Ubuntu 20.04 LTS)
CPU(s):         1
Load:           0.00 0.00 0.00
Disk usage:     1.6GiB out of 48.4GiB
Memory usage:   145.5MiB out of 7.8GiB
Mounts:         --
```

(This also seems weird, because I requested "noble" = 24.04)

- `multipass get local.driver`: qemu

**Additional context**

Following https://documentation.ubuntu.com/public-images/en/latest/public-images-how-to/build-vagrant-with-bartender/ I built multipass from source (because I don't want to use snap). The build failed twice, by the way, and I had to `sudo apt install ninja-build clang` to fix it - would be nice to add those to the README.

I then ran multipassd as root in one terminal window, and in another ran:

```
UBUNTU_SERIES=noble
./ubuntu-bartender --livecd-rootfs-branch ubuntu/$UBUNTU_SERIES --build-provider multipass -- --series $UBUNTU_SERIES --image-target vagrant --project ubuntu-cpc
```

This failed:

```
Launched: noted-ant-ubuntu-bartender
Preparing ingredients... in /home/em/ubuntu-bartender-multipass.FryRQJvSQB
UBUNTU_OLD_FASHIONED_DIR is not set, cloning https://github.com/ubuntu-bartenders/ubuntu-old-fashioned.git to /home/em/ubuntu-bartender-multipass.FryRQJvSQB/ubuntu-old-fashioned ...
LIVECD_ROOTFS_DIR is not set, cloning https://git.launchpad.net/livecd-rootfs to /home/em/ubuntu-bartender-multipass.FryRQJvSQB/livecd-rootfs ...
uploading ingredients.tar.gz to noted-ant-ubuntu-bartender ...
extracting ingredients.tar.gz in noted-ant-ubuntu-bartender ...

gzip: stdin: invalid compressed data--crc error

gzip: stdin: invalid compressed data--length error
tar: Unexpected EOF in archive
tar: Unexpected EOF in archive
tar: Error is not recoverable: exiting now
```

To investigate further I opened a shell with `multipass exec bash` and saw that the ingredients.tar.gz file in the VM is about half the size of the original file on the host. I then reproduced this by uploading the file with `multipass transfer` myself (without bartender).

Tried to build the release/1.15 branch instead of HEAD and `cmake` failed with

```
CMake Error at CMakeLists.txt:162 (message):
  You need to set MULTIPASS_UPSTREAM for a release build.

  Use an empty string to make local the authoritative repository.
Call Stack (most recent call first):
  CMakeLists.txt:216 (determine_version)
```

To a user, who is just trying to build multipass to run it, this message is unclear. How do I set MULTIPASS_UPSTREAM? I tried `export MULTIPASS_UPSTREAM=` but still got the same error. And what does "make local the authoritative repository" mean? Do I want to do that?

Searched PRs here and found that `cmake ../ -DMULTIPASS_UPSTREAM=` makes it work, but it would be nice to improve the message to something like: "If you're just building the release for personal use set this to an empty string by running cmake with -DMULTIPASS_UPSTREAM=".

--- a/src/ssh/sftp_client.cpp
+++ b/src/ssh/sftp_client.cpp
@@ -330,10 +330,12 @@ void SFTPClient::do_push_file(std::istream& source, const fs::path& target_path)
     if (!remote_file)
         throw SFTPError{"cannot open remote file {}: {}", target_path, ssh_get_error(sftp->session)};
 
+    // create an uninitialized buffer to use.
     const auto max_write = sftp_limits(sftp.get())->max_write_length;
-    std::vector<char> buffer(max_write);
-    while (auto r = source.read(buffer.data(), buffer.size()).gcount())
-        if (sftp_write(remote_file.get(), buffer.data(), r) < 0)
+    const std::unique_ptr<char[]> buffer{new char[max_write]};
+
+    while (auto r = source.read(buffer.get(), max_write).gcount())
+        if (sftp_write(remote_file.get(), buffer.get(), r) < 0)
             throw SFTPError{"cannot write to remote file {}: {}", target_path, ssh_get_error(sftp->session)};
 }
 
@@ -343,14 +345,16 @@ void SFTPClient::do_pull_file(const fs::path& source_path, std::ostream& target)
     if (!remote_file)
         throw SFTPError{"cannot open remote file {}: {}", source_path, ssh_get_error(sftp->session)};
 
+    // create an uninitialized buffer to use.
     const auto max_read = sftp_limits(sftp.get())->max_read_length;
-    std::vector<char> buffer(max_read);
-    while (auto r = sftp_read(remote_file.get(), buffer.data(), buffer.size()))
+    const std::unique_ptr<char[]> buffer{new char[max_read]};
+
+    while (auto r = sftp_read(remote_file.get(), buffer.get(), max_read))
     {
         if (r < 0)
             throw SFTPError{"cannot read from remote file {}: {}", source_path, ssh_get_error(sftp->session)};
 
-        target.write(buffer.data(), r);
+        target.write(buffer.get(), r);
     }
 }
 

--- a/src/ssh/sftp_client.cpp
+++ b/src/ssh/sftp_client.cpp
@@ -30,7 +30,6 @@
 #include <fmt/std.h>
 
 constexpr int file_mode = 0664;
-constexpr auto max_transfer = 65536u;
 const std::string stream_file_name{"stream_output.dat"};
 const char* log_category = "sftp";
 
@@ -331,7 +330,8 @@ void SFTPClient::do_push_file(std::istream& source, const fs::path& target_path)
     if (!remote_file)
         throw SFTPError{"cannot open remote file {}: {}", target_path, ssh_get_error(sftp->session)};
 
-    std::array<char, max_transfer> buffer{};
+    const auto max_write = sftp_limits(sftp.get())->max_write_length;
+    std::vector<char> buffer(max_write);
     while (auto r = source.read(buffer.data(), buffer.size()).gcount())
         if (sftp_write(remote_file.get(), buffer.data(), r) < 0)
             throw SFTPError{"cannot write to remote file {}: {}", target_path, ssh_get_error(sftp->session)};
@@ -343,7 +343,8 @@ void SFTPClient::do_pull_file(const fs::path& source_path, std::ostream& target)
     if (!remote_file)
         throw SFTPError{"cannot open remote file {}: {}", source_path, ssh_get_error(sftp->session)};
 
-    std::array<char, max_transfer> buffer{};
+    const auto max_read = sftp_limits(sftp.get())->max_read_length;
+    std::vector<char> buffer(max_read);
     while (auto r = sftp_read(remote_file.get(), buffer.data(), buffer.size()))
     {
         if (r < 0)

--- a/tests/test_sftp_client.cpp
+++ b/tests/test_sftp_client.cpp
@@ -81,6 +81,8 @@ struct SFTPClient : public testing::Test
     MockScope<decltype(mock_sftp_new)> sftp_new;
     MockScope<decltype(mock_sftp_free)> free_sftp;
 
+    sftp_limits_struct limits{32768, 32768, 32768, 0};
+
     const mpt::StubSSHKeyProvider key_provider;
     mpt::MockSSHTestFixture mock_ssh_test_fixture;
 
@@ -115,7 +117,10 @@ TEST_F(SFTPClient, throws_when_failed_to_init)
 
 TEST_F(SFTPClient, is_dir)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
 
     auto mocked_sftp_stat = MOCK(sftp_stat);
     auto sftp_client = make_sftp_client();
@@ -132,7 +137,10 @@ TEST_F(SFTPClient, push_file_success)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _))
@@ -159,7 +167,10 @@ TEST_F(SFTPClient, push_file_success)
 
 TEST_F(SFTPClient, push_file_cannot_open_source)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     auto err = EACCES;
@@ -179,7 +190,10 @@ TEST_F(SFTPClient, push_file_cannot_open_source)
 
 TEST_F(SFTPClient, push_file_cannot_open_target)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
@@ -197,7 +211,10 @@ TEST_F(SFTPClient, push_file_cannot_write_target)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _))
@@ -218,7 +235,10 @@ TEST_F(SFTPClient, push_file_cannot_read_source)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -248,7 +268,10 @@ TEST_F(SFTPClient, push_file_cannot_set_perms)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _))
@@ -274,7 +297,10 @@ TEST_F(SFTPClient, pull_file_success)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
     std::stringstream test_file;
@@ -307,7 +333,10 @@ TEST_F(SFTPClient, pull_file_success)
 
 TEST_F(SFTPClient, pull_file_cannot_open_source)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(); });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_write(target_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
@@ -323,7 +352,10 @@ TEST_F(SFTPClient, pull_file_cannot_open_source)
 
 TEST_F(SFTPClient, pull_file_cannot_open_target)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(); });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     auto err = EACCES;
@@ -343,7 +375,10 @@ TEST_F(SFTPClient, pull_file_cannot_open_target)
 
 TEST_F(SFTPClient, pull_file_cannot_write_target)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
     auto test_file = std::make_unique<std::stringstream>();
@@ -372,7 +407,10 @@ TEST_F(SFTPClient, pull_file_cannot_write_target)
 
 TEST_F(SFTPClient, pull_file_cannot_read_source)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(); });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_write(target_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
@@ -390,7 +428,10 @@ TEST_F(SFTPClient, pull_file_cannot_read_source)
 
 TEST_F(SFTPClient, pull_file_cannot_set_perms)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_write(target_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
     REPLACE(sftp_open, [](auto sftp, auto...) { return get_dummy_sftp_file(sftp); });
@@ -410,7 +451,10 @@ TEST_F(SFTPClient, pull_file_cannot_set_perms)
 
 TEST_F(SFTPClient, push_dir_success_regular)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -449,7 +493,10 @@ TEST_F(SFTPClient, push_dir_success_regular)
 
 TEST_F(SFTPClient, push_dir_success_dir)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -480,7 +527,10 @@ TEST_F(SFTPClient, push_dir_success_dir)
 
 TEST_F(SFTPClient, push_dir_fail_dir)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -513,7 +563,10 @@ TEST_F(SFTPClient, push_dir_fail_dir)
 
 TEST_F(SFTPClient, push_dir_success_symlink)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -543,7 +596,10 @@ TEST_F(SFTPClient, push_dir_success_symlink)
 
 TEST_F(SFTPClient, push_dir_cannot_read_symlink)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -576,7 +632,10 @@ TEST_F(SFTPClient, push_dir_cannot_read_symlink)
 
 TEST_F(SFTPClient, push_dir_cannot_create_symlink)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -610,7 +669,10 @@ TEST_F(SFTPClient, push_dir_cannot_create_symlink)
 
 TEST_F(SFTPClient, push_dir_symlink_over_dir)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -641,7 +703,10 @@ TEST_F(SFTPClient, push_dir_symlink_over_dir)
 
 TEST_F(SFTPClient, push_dir_unknown_file_type)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -668,7 +733,10 @@ TEST_F(SFTPClient, push_dir_unknown_file_type)
 
 TEST_F(SFTPClient, push_dir_open_iter_fail)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -687,7 +755,10 @@ TEST_F(SFTPClient, push_dir_open_iter_fail)
 
 TEST_F(SFTPClient, push_dir_cannot_access_target)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
 
     auto err = std::make_error_code(std::errc::permission_denied);
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce([&](auto, std::error_code& e) {
@@ -703,7 +774,10 @@ TEST_F(SFTPClient, push_dir_cannot_access_target)
 
 TEST_F(SFTPClient, push_dir_r_not_specified)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
 
     auto sftp_client = make_sftp_client();
@@ -715,7 +789,10 @@ TEST_F(SFTPClient, push_dir_r_not_specified)
 
 TEST_F(SFTPClient, pull_dir_success_regular)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
     auto iter = std::make_unique<mpt::MockSFTPDirIterator>();
@@ -767,7 +844,10 @@ TEST_F(SFTPClient, pull_dir_success_regular)
 
 TEST_F(SFTPClient, pull_dir_success_dir)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -788,7 +868,10 @@ TEST_F(SFTPClient, pull_dir_success_dir)
 
 TEST_F(SFTPClient, pull_dir_fail_dir)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -818,7 +901,10 @@ TEST_F(SFTPClient, pull_dir_fail_dir)
 
 TEST_F(SFTPClient, pull_dir_success_symlink)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -843,7 +929,10 @@ TEST_F(SFTPClient, pull_dir_success_symlink)
 
 TEST_F(SFTPClient, pull_dir_cannot_read_symlink)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -869,7 +958,10 @@ TEST_F(SFTPClient, pull_dir_cannot_read_symlink)
 
 TEST_F(SFTPClient, pull_dir_cannot_create_symlink)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -898,7 +990,10 @@ TEST_F(SFTPClient, pull_dir_cannot_create_symlink)
 
 TEST_F(SFTPClient, pull_dir_symlink_over_dir)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -923,7 +1018,10 @@ TEST_F(SFTPClient, pull_dir_symlink_over_dir)
 
 TEST_F(SFTPClient, pull_dir_unknown_file_type)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -945,7 +1043,10 @@ TEST_F(SFTPClient, pull_dir_unknown_file_type)
 
 TEST_F(SFTPClient, pull_dir_r_not_specified)
 {
-    REPLACE(sftp_init, [](auto...) { return SSH_OK; });
+    REPLACE(sftp_init, [this](sftp_session sftp) {
+        sftp->limits = &limits;
+        return SSH_OK;
+    });
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
 
     auto sftp_client = make_sftp_client();

--- a/tests/test_sftp_client.cpp
+++ b/tests/test_sftp_client.cpp
@@ -77,6 +77,13 @@ struct SFTPClient : public testing::Test
         return {std::make_unique<mp::SSHSession>("b", 43, "ubuntu", key_provider)};
     }
 
+// this is a macro since REPLACE only applies to the current scope and cannot be moved out.
+#define REPLACE_SFTP_INIT()                                                                                            \
+    REPLACE(sftp_init, [this](sftp_session sftp) {                                                                     \
+        sftp->limits = &limits;                                                                                        \
+        return SSH_OK;                                                                                                 \
+    });
+
     decltype(MOCK(sftp_close)) close{MOCK(sftp_close)};
     MockScope<decltype(mock_sftp_new)> sftp_new;
     MockScope<decltype(mock_sftp_free)> free_sftp;
@@ -117,10 +124,7 @@ TEST_F(SFTPClient, throws_when_failed_to_init)
 
 TEST_F(SFTPClient, is_dir)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
 
     auto mocked_sftp_stat = MOCK(sftp_stat);
     auto sftp_client = make_sftp_client();
@@ -137,10 +141,7 @@ TEST_F(SFTPClient, push_file_success)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _))
@@ -167,10 +168,7 @@ TEST_F(SFTPClient, push_file_success)
 
 TEST_F(SFTPClient, push_file_cannot_open_source)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     auto err = EACCES;
@@ -190,10 +188,7 @@ TEST_F(SFTPClient, push_file_cannot_open_source)
 
 TEST_F(SFTPClient, push_file_cannot_open_target)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
@@ -211,10 +206,7 @@ TEST_F(SFTPClient, push_file_cannot_write_target)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _))
@@ -235,10 +227,7 @@ TEST_F(SFTPClient, push_file_cannot_read_source)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -268,10 +257,7 @@ TEST_F(SFTPClient, push_file_cannot_set_perms)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(false));
     EXPECT_CALL(*mock_sftp_utils, get_remote_file_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_read(source_path, _))
@@ -297,10 +283,7 @@ TEST_F(SFTPClient, pull_file_success)
 {
     std::string test_data = "test_data";
 
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
     std::stringstream test_file;
@@ -333,10 +316,7 @@ TEST_F(SFTPClient, pull_file_success)
 
 TEST_F(SFTPClient, pull_file_cannot_open_source)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(); });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_write(target_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
@@ -352,10 +332,7 @@ TEST_F(SFTPClient, pull_file_cannot_open_source)
 
 TEST_F(SFTPClient, pull_file_cannot_open_target)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(); });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     auto err = EACCES;
@@ -375,10 +352,7 @@ TEST_F(SFTPClient, pull_file_cannot_open_target)
 
 TEST_F(SFTPClient, pull_file_cannot_write_target)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
     auto test_file = std::make_unique<std::stringstream>();
@@ -407,10 +381,7 @@ TEST_F(SFTPClient, pull_file_cannot_write_target)
 
 TEST_F(SFTPClient, pull_file_cannot_read_source)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(); });
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_write(target_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
@@ -428,10 +399,7 @@ TEST_F(SFTPClient, pull_file_cannot_read_source)
 
 TEST_F(SFTPClient, pull_file_cannot_set_perms)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_sftp_utils, get_local_file_target(source_path, target_path, _)).WillOnce(Return(target_path));
     EXPECT_CALL(*mock_file_ops, open_write(target_path, _)).WillOnce(Return(std::make_unique<std::stringstream>()));
     REPLACE(sftp_open, [](auto sftp, auto...) { return get_dummy_sftp_file(sftp); });
@@ -451,10 +419,7 @@ TEST_F(SFTPClient, pull_file_cannot_set_perms)
 
 TEST_F(SFTPClient, push_dir_success_regular)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -493,10 +458,7 @@ TEST_F(SFTPClient, push_dir_success_regular)
 
 TEST_F(SFTPClient, push_dir_success_dir)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -527,10 +489,7 @@ TEST_F(SFTPClient, push_dir_success_dir)
 
 TEST_F(SFTPClient, push_dir_fail_dir)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -563,10 +522,7 @@ TEST_F(SFTPClient, push_dir_fail_dir)
 
 TEST_F(SFTPClient, push_dir_success_symlink)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -596,10 +552,7 @@ TEST_F(SFTPClient, push_dir_success_symlink)
 
 TEST_F(SFTPClient, push_dir_cannot_read_symlink)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -632,10 +585,7 @@ TEST_F(SFTPClient, push_dir_cannot_read_symlink)
 
 TEST_F(SFTPClient, push_dir_cannot_create_symlink)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -669,10 +619,7 @@ TEST_F(SFTPClient, push_dir_cannot_create_symlink)
 
 TEST_F(SFTPClient, push_dir_symlink_over_dir)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -703,10 +650,7 @@ TEST_F(SFTPClient, push_dir_symlink_over_dir)
 
 TEST_F(SFTPClient, push_dir_unknown_file_type)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -733,10 +677,7 @@ TEST_F(SFTPClient, push_dir_unknown_file_type)
 
 TEST_F(SFTPClient, push_dir_open_iter_fail)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
     EXPECT_CALL(*mock_sftp_utils, get_remote_dir_target(_, source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -755,10 +696,7 @@ TEST_F(SFTPClient, push_dir_open_iter_fail)
 
 TEST_F(SFTPClient, push_dir_cannot_access_target)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
 
     auto err = std::make_error_code(std::errc::permission_denied);
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce([&](auto, std::error_code& e) {
@@ -774,10 +712,7 @@ TEST_F(SFTPClient, push_dir_cannot_access_target)
 
 TEST_F(SFTPClient, push_dir_r_not_specified)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_file_ops, is_directory(source_path, _)).WillOnce(Return(true));
 
     auto sftp_client = make_sftp_client();
@@ -789,10 +724,7 @@ TEST_F(SFTPClient, push_dir_r_not_specified)
 
 TEST_F(SFTPClient, pull_dir_success_regular)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
     auto iter = std::make_unique<mpt::MockSFTPDirIterator>();
@@ -844,10 +776,7 @@ TEST_F(SFTPClient, pull_dir_success_regular)
 
 TEST_F(SFTPClient, pull_dir_success_dir)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -868,10 +797,7 @@ TEST_F(SFTPClient, pull_dir_success_dir)
 
 TEST_F(SFTPClient, pull_dir_fail_dir)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -901,10 +827,7 @@ TEST_F(SFTPClient, pull_dir_fail_dir)
 
 TEST_F(SFTPClient, pull_dir_success_symlink)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -929,10 +852,7 @@ TEST_F(SFTPClient, pull_dir_success_symlink)
 
 TEST_F(SFTPClient, pull_dir_cannot_read_symlink)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -958,10 +878,7 @@ TEST_F(SFTPClient, pull_dir_cannot_read_symlink)
 
 TEST_F(SFTPClient, pull_dir_cannot_create_symlink)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -990,10 +907,7 @@ TEST_F(SFTPClient, pull_dir_cannot_create_symlink)
 
 TEST_F(SFTPClient, pull_dir_symlink_over_dir)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -1018,10 +932,7 @@ TEST_F(SFTPClient, pull_dir_symlink_over_dir)
 
 TEST_F(SFTPClient, pull_dir_unknown_file_type)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
     EXPECT_CALL(*mock_sftp_utils, get_local_dir_target(source_path, target_path, _)).WillOnce(Return(target_path));
 
@@ -1043,10 +954,7 @@ TEST_F(SFTPClient, pull_dir_unknown_file_type)
 
 TEST_F(SFTPClient, pull_dir_r_not_specified)
 {
-    REPLACE(sftp_init, [this](sftp_session sftp) {
-        sftp->limits = &limits;
-        return SSH_OK;
-    });
+    REPLACE_SFTP_INIT();
     REPLACE(sftp_stat, [](auto...) { return get_dummy_sftp_attr(SSH_FILEXFER_TYPE_DIRECTORY); });
 
     auto sftp_client = make_sftp_client();

